keps/0018-controller-overhaul.md

```diff
@@ -0,0 +1,121 @@
+      * [Proposal](#proposal)
+         * [Instance CRD changes](#instance-crd-changes)
+         * [Admission webhook](#admission-webhook)
+         * [Implementation details](#implementation-detailsnotesconstraints)
+         * [Risks and mitigations](#risks-and-mitigations)
+      * [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+The way KUDO controllers are implemented currently has several flaws:
+1. When KUDO manager is down (or restarted) we might miss an update on a CRD meaning we won’t execute the plan that was supposed to run at or we execute incorrect one ([issue](https://github.com/kudobuilder/kudo/issues/422))
+2. Multiple plan can be seemingly in progress leading to misleading status communicated ([issue](https://github.com/kudobuilder/kudo/issues/628))
+3. No way to ensure atomicity on plan execution (given the current CRD design where information is spread across several CRDs)
+4. Very low test coverage and overall confidence in the code
+
+To address all of the above points we’re proposing new design of our CRDs. We’re going to get rid of PlanExecution CRD altogether and merge the execution status into the Instance’s status subresource. At the same time we’re going to keep last known state of instance inside status subresource as well, which will allow us to properly identify plans that are supposed to be run after user submits a change to the Instance spec.
+
+## Motivation
+
+### Goals
+
+- better understandable and tested controller code
+- ensuring correct plan is executed in case KUDO controller is not running for some time
+- atomicity on plan execution level - do not allow to run plan when other plan is in progress
+- be able to see current plan execution status
+
+### Non-Goals
+
+- providing longer history of plan executions that just the last one run
+- have manual plan execution feature (there are several approaches to how to do this, I think it deserves a separate KEP)
+
+## Proposal
+
+The proposed design that will solve all the current known problems with controllers and CRDs is to get rid of PlanExecution CRD altogether. Plan status will basically become instance status (it’s the case right now as well anyway, we just need to replicate those values across two places). 
+
+### Instance CRD changes
+
+The instance CRD for instance with running deploy plan might then look as following.
+
+```yaml
+ Instance
+  spec:
+    operatorVersion: operator-123
+    parameters:
+      - REPLICAS: 4
+  status:
+    planStatus: # (for every plan defined in OV, there will be one field)
+      - deploy:
+          status: IN_PROGRESS
+          executionName: deploy-1478631057
+          lastFinishedRun: 1568642057
+          overview:
+            strategy: serial
+            phases:
+              - name: deploy
+                strategy: serial
+                state: IN_PROGRESS
+                steps:
+                  - name: everything
+                    state: IN_PROGRESS
+                    delete: false
```

Comment:
Why is `delete` needed here? Is it part of the step status?

Reply:
hmm I think it actually is not. Delete is a property of the step but if that step is a delete step is already present in the plan definition. I think this one can go. Good catch